### PR TITLE
[Feature] Add GradCAM utils for Recognizer.

### PR DESCRIPTION
**Pull request by irvingzhang0512**
+ [x] add unittest
+ [x] test generating heatmaps with labels.

### CODE REVIEW

File: tests/test_models/test_recognizers.py [hunk 1]
```diff
@@ -26,6 +26,9 @@ def forward_train(self, imgs, labels):
     def forward_test(self, imgs):
         pass
 
+    def forward_gradcam(self, imgs):
```
Comment (SuX97) [inline tests/test_models/test_recognizers.py:29]: Need unittest.

File: mmaction/utils/gradcam_utils.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        else:
+            # [batch_size, num_crops*num_clips, 3, clip_len, H, W]
+            b1, b2, _, T, H, W = inputs['imgs'].size()
+            B = b1 * b2
+
+        gradients = self.target_gradients
+        activations = self.target_activations
+        if self.is_recognizer2d:
+            # [B*Tg, C', H', W']
+            BB, C, _, _ = gradients.size()
```
Comment (SuX97) [inline mmaction/utils/gradcam_utils.py:102]: Better use meaning full abbreviate var name.

File: demo/demo_gradcam.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+    # build the recognizer from a config file and checkpoint file
+    model = init_recognizer(
+        args.config,
+        args.checkpoint,
+        device=device,
+        use_frames=args.use_frames)
+
+    inputs = build_inputs(model, args.video, use_frames=args.use_frames)
+    gradcam = GradCAM(model, args.target_layer_name)
```
Comment (SuX97) [inline demo/demo_gradcam.py:100]: Better wrap the following part in a separate function for unittest.

File: demo/demo_gradcam.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    results = gradcam(inputs)
+
+    if args.out_filename is not None:
+        from moviepy.editor import ImageSequenceClip
```
Comment (SuX97) [inline demo/demo_gradcam.py:104]: Is this package necessary? Please add it to requirements/optional.txt.

File: mmaction/models/recognizers/recognizer3d.py [hunk 5]
```diff
@@ -54,3 +54,15 @@ def forward_dummy(self, imgs):
         x = self.extract_feat(imgs)
         outs = (self.cls_head(x), )
         return outs
+
+    def forward_gradcam(self, imgs):
+        """"""
```
Comment (SuX97) [inline mmaction/models/recognizers/recognizer3d.py:59]: A simple but valid docstring is needed.

File: mmaction/utils/gradcam_utils.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+            raise ValueError('GradCAM utils only support Recognizer models.')
+
+        self.target_gradients = None
+        self.target_activations = None
+        self.colormap = plt.get_cmap(colormap)
+        self.data_mean = torch.tensor(model.cfg.img_norm_cfg['mean'])
+        self.data_std = torch.tensor(model.cfg.img_norm_cfg['std'])
+        self._register_single_hook(target_layer_name)
+
+    def _register_single_hook(self, layer_name):
```
Comment (SuX97) [inline mmaction/utils/gradcam_utils.py:39]: Why `single`? Seems like `_register_hooks`.

File: mmaction/utils/gradcam_utils.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        Returns:
+        """
+        imgs = inputs['imgs']
+
+        # localization_map shape [B, T, H, W]
+        # 2D: B = batch_size, T = num_segments
+        # 3D: B = batch_size * num_crops * num_clips, T = clip_len
+        localization_map, preds = self._calculate_localization_map(
+            inputs, use_labels=use_labels)
+        if localization_map.device != torch.device('cpu'):
```
Comment (SuX97) [inline mmaction/utils/gradcam_utils.py:161]: Seems that localization_map should eventually be on cpu, maybe just avoid this if-branch for better unittest coverage.

**Comment by SuX97** [comment]
Cool.

File: demo/demo_gradcam.py (hunk 4 above)
Comment (irvingzhang0512) [reply demo/demo_gradcam.py:104]: Actually, this follows [demo/demo.py](https://github.com/open-mmlab/mmaction2/blob/f8ea6716ca76b2ceef16360b8867ee8e2c1fd82b/demo/demo.py#L92)

**Comment by irvingzhang0512** [comment]
@SuX97 
I'm not familiar with unittest...
It seems that `test_bmn` fails, but I cannot find why.
Any suggestions?

File: tests/test_gradcam.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+
+
+def test_tsn():
+    config = _get_cfg('tsn/tsn_r50_1x1x3_100e_kinetics400_rgb.py')
+    config.model['backbone']['pretrained'] = None
+    recognizer = build_recognizer(config.model, test_cfg=config.test_cfg)
+    recognizer.cfg = config
+
+    # 25(num_clips) * 10(TenCrop) = 250
+    input_shape = (1, 250, 3, 32, 32)
```
Comment (dreamerlin) [inline tests/test_gradcam.py:151]: The shape is too huge for unittest in CI

**Comment by dreamerlin** [comment]
> @SuX97
> I'm not familiar with unittest...
> It seems that `test_bmn` fails, but I cannot find why.
> Any suggestions?

Some data shapes are too huge, which may cause failure in CI.

File: mmaction/utils/gradcam_utils.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+
+        Args:
+            inputs (dict): model inputs, at least including two keys,
+                `imgs` and `label`.
+            use_labels (bool): Whether to use given labels to generate
+                localization map. Labels are in `inputs['label']`.
+        Returns:
+            localization_map (ndarray(s)): the localization map for input imgs.
+            preds (tensor): Model predictions for `inputs`.
+                shape (n_instances, n_class).
```
Comment (dreamerlin) [inline mmaction/utils/gradcam_utils.py:74]:
```text
Returns:
    tuple[np.ndarray, torch.Tensor]: (localization_map, preds)
        localization_map (np.ndarray): the localization map for input imgs.
        preds (torch.Tensor): Model predictions for `inputs` with shape (n_instances, n_class).
```

File: mmaction/utils/gradcam_utils.py [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+        Args:
+            localization_map (tensor): localization map for all inputs,
+                generated with Grad-CAM
+            input_imgs (tensor): model inputs, normed images.
+            alpha (float): transparency level of the heatmap,
+                in the range [0, 1].
+        Returns:
+            blended_imgs (tensor): blending results for localization map and
+                input images. shape [B, T, H, W, 3], pixel values in
+                range [0, 1], in RGB order.
```
Comment (dreamerlin) [inline mmaction/utils/gradcam_utils.py:158]:
```text
torch.Tensor: blending results for localization map and input images. 
    with shape [B, T, H, W, 3] and pixel values in RGB order within range [0, 1].
```

File: tests/test_gradcam.py [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+        ]
+
+    return blended_imgs_target_shape, preds_target_shape
+
+
+def _generate_gradcam_inputs(input_shape=(1, 3, 3, 224, 224), model_type='2D'):
+    """Create a superset of inputs needed to run gradcam.
+
+    Args:
+        input_shape (tuple): input batch dimensions.
```
Comment (dreamerlin) [inline tests/test_gradcam.py:61]: `tuple[int]`

File: tests/test_gradcam.py [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+    blended_imgs_target_shape, preds_target_shape = _get_target_shapes(
+        input_shape, num_classes=num_classes, model_type='2D')
+
+    blended_imgs, preds = gradcam(demo_inputs)
+    assert list(blended_imgs.size()) == blended_imgs_target_shape
+    assert list(preds.size()) == preds_target_shape
+
+    blended_imgs, preds = gradcam(demo_inputs, True)
+    assert list(blended_imgs.size()) == blended_imgs_target_shape
+    assert list(preds.size()) == preds_target_shape
```
[the remark this answers is not in the document]
Comment (irvingzhang0512) [reply tests/test_gradcam.py:106]: `type(tensor.size())` returns `torch.Size`, not `list`

```python
a = torch.rand(1, 2)
print(a.size() == [1, 2]) # False
```

**Comment by dreamerlin** [comment]
The docs about demo can be moved to [demo/README.md](https://github.com/open-mmlab/mmaction2/blob/master/demo/README.md)

File: tests/test_gradcam.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (dreamerlin) [reply tests/test_gradcam.py:106]: set shape as tuple
```python
a = torch.rand(1, 2)
print(a.size() == (1, 2)) # True
```

[the remark this answers is not in the document]
Comment (irvingzhang0512) [reply tests/test_gradcam.py:106]: That's good to know. I'll convert all shape to tuple.

File: mmaction/utils/gradcam_utils.py [hunk 13]
```diff
@@ -0,0 +1,230 @@
+import matplotlib.pyplot as plt
```
Comment (dreamerlin) [inline mmaction/utils/gradcam_utils.py:1]: put the `gradcam_utils` in `mmaction/utils/__init__.py`

```python
...
from .gradcam_utils import GradCAM
...

__all__ = [..., 'GradCAM']
```

File: demo/demo_gradcam.py [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+
+import mmcv
+import numpy as np
+import torch
+from mmcv.parallel import collate, scatter
+from moviepy.editor import ImageSequenceClip
+
+from mmaction.apis import init_recognizer
+from mmaction.datasets.pipelines import Compose
+from mmaction.utils.gradcam_utils import GradCAM
```
Comment (dreamerlin) [inline demo/demo_gradcam.py:13]: changed to `from mmaction.utils import GradCAM`, when set it in `mmaction/utils/__init__.py`

File: mmaction/utils/gradcam_utils.py (hunk 13 above)
Comment (innerlee) [inline mmaction/utils/gradcam_utils.py:1]: can plt be avoided here?

File: demo/demo_gradcam.py [hunk 15]
```diff
@@ -0,0 +1,196 @@
+import argparse
+import os
+import os.path as osp
+
+import mmcv
+import numpy as np
+import torch
+from mmcv.parallel import collate, scatter
+from moviepy.editor import ImageSequenceClip
```
Comment (innerlee) [inline demo/demo_gradcam.py:9]: can moviepy be replaced?

**Comment by dreamerlin** [comment]
Please also help to highlight that the demo can also loads checkpoint from corresponding url, like [this](https://github.com/open-mmlab/mmaction2/pull/339). thanks!

File: mmaction/utils/gradcam_utils.py [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+        localization_map_min, localization_map_max = (
+            torch.min(localization_map.view(b, -1), dim=-1, keepdim=True)[0],
+            torch.max(localization_map.view(b, -1), dim=-1, keepdim=True)[0])
+        localization_map_min = torch.reshape(
+            localization_map_min, shape=(b, 1, 1, 1, 1))
+        localization_map_max = torch.reshape(
+            localization_map_max, shape=(b, 1, 1, 1, 1))
+        # delta must be small enough.
+        # make sure `localization_map_max - localization_map_min >> delta`
+        delta = 1e-20
```
Comment (dreamerlin) [inline mmaction/utils/gradcam_utils.py:145]: It is not good to use hardcoded value for `delta`. you can set it as keyword argument in the function.

```python
def _calculate_localization_map(self, inputs, use_labels, delta=1e-20):
```

And do not forget to update the docstring, thx : )

File: demo/demo_gradcam.py (hunk 14 above)
Comment (irvingzhang0512) [reply demo/demo_gradcam.py:13]: After updating codes after your suggestion, `ImportError` occurs...
Maybe `from ..models.recognizers import Recognizer2D, Recognizer3D` causes some recurrent import?

Comment (dreamerlin) [reply demo/demo_gradcam.py:12]: try to put `from ..models.recognizers import Recognizer2D, Recognizer3D` into 'def __init__()' of `class GradCAM`

File: mmaction/utils/gradcam_utils.py [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+
+        Args:
+            model (nn.Module): the recognizer model to be used.
+            target_layer_name (str): name of convolutional layer to
+                be used to get gradients and feature maps from for creating
+                localization maps.
+            colormap (Optional[str]): matplotlib colormap used to create
+                heatmap. For more information, please visit:
+                https://matplotlib.org/3.3.0/tutorials/colors/colormaps.html
+        """
```
Comment (dreamerlin) [inline mmaction/utils/gradcam_utils.py:27]: Move `from ..models.recognizers import Recognizer2D, Recognizer3D` here.

**Comment by dreamerlin** [comment]
LGTM

**Review by dreamerlin — APPROVED** [review]

File: mmaction/models/recognizers/base.py [hunk 18]
```diff
@@ -154,9 +160,16 @@ def _parse_losses(losses):
 
         return loss, log_vars
 
-    def forward(self, imgs, label=None, return_loss=True, **kwargs):
+    def forward(self,
+                imgs,
+                label=None,
+                gradcam=False,
```
Comment (SuX97) [inline mmaction/models/recognizers/base.py:166]: I prefer not to add too many custom features in BaseClass, maybe catch `gradcam` by:
```python
        # if gradcam:
        if kwargs.get(gradcam, False)
            return self.forward_gradcam(imgs, **kwargs)
```
Since we already allow kwargs.

**Review by SuX97 — APPROVED** [review]
LGTM.

**Comment by innerlee** [comment]
This is a big pr, thanks!

**Comment by innerlee** [comment]
good to go after resolving the conflicts

**Comment by dreamerlin** [comment]
> good to go after resolving the conflicts

Help with resolving conflicts
